Mathlib/RingTheory/Valuation/Basic.lean

```diff
@@ -472,29 +474,442 @@ theorem map {v' : Valuation R Γ₀} (f : Γ₀ →*₀ Γ'₀) (hf : Monotone f
+    obtain le | le := le_total (v₁ r₁) (v₁ r₂)
+    · have le₂ := (h _ _).mp le
+      exact ⟨r₁, by simp only [min_eq_left le, min_eq_left le₂, true_and]⟩
+    · have le₂ := (h _ _).mp le
+      exact ⟨r₂, by simp only [min_eq_right le, min_eq_right le₂, true_and]⟩
+  case max hxy hzw ih₁ ih₂ =>
+    obtain ⟨⟨r₁, rfl, rfl⟩, ⟨r₂, rfl, rfl⟩⟩ := ih₁, ih₂
+    obtain le | le := le_total (v₁ r₁) (v₁ r₂)
+    · have le₂ := (h _ _).mp le
+      exact ⟨r₂, by simp only [max_eq_right le, max_eq_right le₂, true_and]⟩
+    · have le₂ := (h _ _).mp le
+      exact ⟨r₁, by simp only [max_eq_left le, max_eq_left le₂, true_and]⟩
+
+variable {x y : Γ₁} {z w : Γ₂} (hxz : h.Associatedₘ x z) (hyw : h.Associatedₘ y w)
+include hxz hyw
+
+theorem le_iff_le : x ≤ y ↔ z ≤ w := by
+  obtain ⟨⟨r₁, rfl, rfl⟩, ⟨r₂, rfl, rfl⟩⟩ := hxz.exists_eq, hyw.exists_eq
+  exact h r₁ r₂
+
+theorem lt_iff_lt : x < y ↔ z < w := by
+  simp_rw [← not_le, hyw.le_iff_le hxz]
+
+theorem eq_iff_eq : x = y ↔ z = w := by
+  simp_rw [le_antisymm_iff, hxz.le_iff_le hyw, hyw.le_iff_le hxz]
+
+theorem ne_iff_ne : x ≠ y ↔ z ≠ w := by
+  simp_rw [not_iff_not, hxz.eq_iff_eq hyw]
+
+end Associatedₘ
+
+end LinearOrderedCommMonoidWithZero
+
+section LinearOrderedCommGroupWithZero
+
+variable [Ring R]
+  {Γ₁ Γ₂ : Type*} [LinearOrderedCommGroupWithZero Γ₁] [LinearOrderedCommGroupWithZero Γ₂]
+  {v₁ : Valuation R Γ₁} {v₂ : Valuation R Γ₂}
+  (h : v₁.IsEquiv v₂)
+  {x x₁ x₂ y y₁ y₂ z w : R}
+
+/-- If `v₁.Equiv v₂`, then the value groups of `v₁` and `v₂` are isomorphic, and this `Prop` is
+precisely the condition that the isomorphism sends `x : Γ₁` to `y : Γ₂`. -/
+inductive Associated (h : v₁.IsEquiv v₂) : Γ₁ → Γ₂ → Prop where
+  | value {z : R} : h.Associated (v₁ z) (v₂ z)
+  | inv {x : Γ₁} {y : Γ₂} (hxy : h.Associated x y) : h.Associated x⁻¹ y⁻¹
+  | mul {x y : Γ₁} {z w : Γ₂} (hxz : h.Associated x z) (hyw : h.Associated y w) :
+      h.Associated (x * y) (z * w)
+  | min {x y : Γ₁} {z w : Γ₂} (hxz : h.Associated x z) (hyw : h.Associated y w) :
+      h.Associated (min x y) (min z w)
+  | max {x y : Γ₁} {z w : Γ₂} (hxz : h.Associated x z) (hyw : h.Associated y w) :
+      h.Associated (max x y) (max z w)
+
+include h
+
+lemma div_le_div_iff_div_le_div : v₁ x / v₁ y ≤ v₁ z / v₁ w ↔ v₂ x / v₂ y ≤ v₂ z / v₂ w := by
+  by_cases hy : v₁ y = 0
+  · rw [hy, h.eq_zero.1 hy]; simp
+  by_cases hw : v₁ w = 0
+  · rw [hw, h.eq_zero.1 hw]; simp [h.eq_zero]
+  have hy₂ : v₂ y ≠ 0 := h.ne_zero.1 hy
+  have hw₂ : v₂ w ≠ 0 := h.ne_zero.1 hw
+  rw [div_le_iff₀ (zero_lt_iff.mpr hy), div_le_iff₀ (zero_lt_iff.mpr hy₂),
+    div_mul_eq_mul_div₀, div_mul_eq_mul_div₀,
+    le_div_iff₀ (zero_lt_iff.mpr hw), le_div_iff₀ (zero_lt_iff.mpr hw₂)]
+  exact Associatedₘ.le_iff_le (h := h) (.mul .value .value) (.mul .value .value)
+
+namespace Associated
+
+variable {h}
+
+theorem exists_eq {x : Γ₁} {y : Γ₂} (hxy : h.Associated x y) :
+    ∃ r₁ r₂, v₁ r₁ / v₁ r₂ = x ∧ v₂ r₁ / v₂ r₂ = y := by
```

Comment:
Should this just be the definition of Associated? Then the inductive definition can be an induction principle.

Reply:
resolved in zulip